doc/ECMP/Hierarchical-ECMP.md

```diff
@@ -0,0 +1,60 @@
+Hierarchical ECMP
+-------------------------------------------------------------------------------
+ Title       | SAI support for hierarchical ecmp
+-------------|-----------------------------------------------------------------
+ Authors     | Jai Kumar, Broadcom Inc.
+ Status      | In review
+ Type        | Standards track
+ Created     | 07/08/2022: Initial Draft
+ SAI-Version | 
+-------------------------------------------------------------------------------
+
+Hierarchical ECMP provides two level route resolution for equal cost multipaths. First level of route resolution will happen via level 1 next hop group. Level 1 nexthop group members may point directly to a next hop or may resolve via 2nd level of next hop group. Level 2 next hop group will always point to a next hop.
+
+### Usecase
+There are two common usecases for hierarchical ECMP
+1. EGP Routes resolving via IGP. This is just IP next hop resoltion via another IP next hop.
+2. Overlay or tunnel nexthops resolving via Level 2 next hop group. This is overlay nexthop resoling via underlay next hop or next hop group.
+
+
+
+![Hierarchical ECMP](../figures/H-ECMP.png "Figure 1: Hierarchical ECMP")
+
+__Figure 1: NH resolution in Hierarchical Next Hop Group__
+
+
+Following new attribute is introduced in next hop group object.
+```
+    /**
+     * @brief Hierarchical next hop group level.
+     * false: level 2, underlay next hop group
+     * true : level 1, overlay or tunnel next hop group
+     *
+     * @type bool
+     * @flags CREATE_AND_SET
+     * @default false
+     * @validonly SAI_NEXT_HOP_GROUP_ATTR_TYPE == SAI_NEXT_HOP_GROUP_TYPE_DYNAMIC_UNORDERED_ECMP
+     */
+    SAI_NEXT_HOP_GROUP_ATTR_LEVEL_1,
+```
+
+Following next hop group member attribute already exists that points the level 2 next hop group object id.
+```
+    /**
+     * @brief Next hop id
+     *
+     * @type sai_object_id_t
+     * @flags MANDATORY_ON_CREATE | CREATE_AND_SET
+     * @objects SAI_OBJECT_TYPE_NEXT_HOP, SAI_OBJECT_TYPE_NEXT_HOP_GROUP
+     */
+    SAI_NEXT_HOP_GROUP_MEMBER_ATTR_NEXT_HOP_ID,
+```
+
+### Sample Workflow
+NOS will set the new attribute SAI_NEXT_HOP_GROU_ATTR_LEVEL_1 to true when next hop group object is for overlay or EGP routes. This group can contain next hop and/or next hop group object id as its member.
+SAI Adapter will allocate overlay nexthop group in the corresponding HW table.
+

```

Comment:
Please give examples of :

1) a nexthop group that has both level1 and level2 nexthops

2) how this would work with MPLS VPNs that can have more hierarchies

Reply:
Change the attribute based on our offline discussions